File: frontend/src/api/resumes.ts
import client from './client';
import type { Resume, Module } from '../types/resume';

export async function listResumes(): Promise<Resume[]> {
  const { data } = await client.get('/resumes');
  return data;
}

export async function getResume(id: number): Promise<Resume> {
  const { data } = await client.get(`/resumes/${id}`);
  return data;
}

export async function createResume(title: string): Promise<Resume> {
  const { data } = await client.post('/resumes', { title });
  return data;
}

export async function updateResume(id: number, body: { title?: string; template_id?: string; font_scale?: number; primary_color?: string; accent_color?: string }): Promise<Resume> {
  const { data } = await client.put(`/resumes/${id}`, body);
  return data;
}

export async function deleteResume(id: number): Promise<void> {
  await client.delete(`/resumes/${id}`);
}

export async function createModule(resumeId: number, type: string): Promise<Module> {
  const { data } = await client.post(`/resumes/${resumeId}/modules`, { type });
  return data;
}

export async function updateModule(id: number, body: { visible?: boolean; data?: string }): Promise<Module> {
  const { data } = await client.put(`/modules/${id}`, body);
  return data;
}

export async function deleteModule(id: number): Promise<void> {
  await client.delete(`/modules/${id}`);
}

export async function reorderModules(resumeId: number, order: number[]): Promise<Module[]> {
  const { data } = await client.put(`/resumes/${resumeId}/modules/reorder`, { order });
  return data;
}

File: backend/handlers/resumes.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"resume-builder/middleware"
	"resume-builder/models"
	"resume-builder/store"

	"github.com/go-chi/chi/v5"
)

func ListResumes(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r)
	store.DB.RLock()
	resumes := store.DB.ListResumes(userID)
	store.DB.RUnlock()

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resumes)
}

func CreateResume(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r)
	var req models.CreateResumeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, `{"error":"invalid request"}`, http.StatusBadRequest)
		return
	}
	if req.Title == "" {
		req.Title = "未命名简历"
	}

	store.DB.Lock()
	resume := store.DB.CreateResume(userID, req.Title)
	store.DB.Unlock()

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(resume)
}

func GetResume(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r)
	idStr := chi.URLParam(r, "id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		http.Error(w, `{"error":"invalid id"}`, http.StatusBadRequest)
		return
	}

	store.DB.RLock()
	resume := store.DB.FindResume(id, userID)
	if resume == nil {
		store.DB.RUnlock()
		http.Error(w, `{"error":"not found"}`, http.StatusNotFound)
		return
	}
	modules := store.DB.ListModules(id)
	store.DB.RUnlock()

	resume.Modules = modules

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resume)
}

func UpdateResume(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r)
	idStr := chi.URLParam(r, "id")
	id, _ := strconv.ParseInt(idStr, 10, 64)

	var req models.UpdateResumeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, `{"error":"invalid request"}`, http.StatusBadRequest)
		return
	}

	store.DB.Lock()
	resume := store.DB.UpdateResume(id, userID, req.Title, req.TemplateID, req.FontScale, req.PrimaryColor, req.AccentColor)
	store.DB.Unlock()

	if resume == nil {
		http.Error(w, `{"error":"not found"}`, http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resume)
}

func DeleteResume(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r)
	idStr := chi.URLParam(r, "id")
	id, _ := strconv.ParseInt(idStr, 10, 64)

	store.DB.Lock()
	store.DB.DeleteResume(id, userID)
	store.DB.Unlock()

	w.WriteHeader(http.StatusNoContent)
}
